### PR TITLE
KYV-743 Added support to save lastValidPurchaseDateTime to order

### PR DESCRIPTION
**Pull request by severikupari1**

### CODE REVIEW

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/model/Order.java [hunk 1]
```diff
@@ -89,4 +92,16 @@ public Order(String orderId, String namespace, String user, LocalDateTime create
         this.createdAt = createdAt;
         this.incrementId = incrementId;
     }
+
+    public Order(String orderId, String namespace, String user, LocalDateTime createdAt, Long incrementId, LocalDateTime lastValidPurchaseDateTime) {
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/model/Order.java:96]: Why did you create new one? Should you do the parsing for date before and add this value with setter before saving?

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java [hunk 2]
```diff
@@ -131,6 +131,17 @@ public Order createByParams(String namespace, String user) {
         return order;
     }
 
+    public Order createByParams(String namespace, String user, LocalDateTime lastValidPurchaseDateTime) {
+        LocalDateTime createdAt = DateTimeUtil.getFormattedDateTime();
+        String orderId = generateOrderId(namespace, user, createdAt);
+        Long incrementId = this.incrementId.generateOrderIncrementId();
+        Order order = new Order(orderId, namespace, user, createdAt, incrementId, lastValidPurchaseDateTime);
+
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java:139]: Could you add lastValidPurchaseDateTime with setter, using new constructor makes code slower to add new fields and it takes time to change many constructors.

Code could look like 
`
       ....
        if (StringUtils.isNotEmpty(lastValidPurchaseDateTime)) {
            order.lastValidPurchaseDateTime(lastValidPurchaseDateTime);
        }
        orderRepository.save(order);
        log.debug("created new order, orderId: " + orderId);
        return order;
`

File: common/src/main/java/fi/hel/verkkokauppa/common/util/DateTimeUtil.java [hunk 3]
```diff
@@ -53,6 +54,15 @@ public static LocalDateTime fromFormattedDateTimeString(String localDateTime) {
         return LocalDateTime.parse(localDateTime, formatter);
     }
 
+    public static LocalDateTime fromFormattedDateTimeOptionalString(String localDateTime) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
```
Comment (severikupari1) [inline common/src/main/java/fi/hel/verkkokauppa/common/util/DateTimeUtil.java:58]: Could that pattern to be moved to constant and given name of the format used. Constant can be created to this class.

DateTimeUtil.DATE_TIME_WITH_MILLISECONDS etc

**Comment by severikupari1** [comment]
This is @synthko :s pr not mine 👍

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/model/Order.java (hunk 1 above)
Comment (synthko) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/model/Order.java:96]: When aiming to get things working it did not cross my mind to just use setters after initialising the order with the old constructor. But yes, that makes much more sense so I'll remove my constructor and use setters to fill in the lastValidPurchaseDateTime for the order. 👍

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java (hunk 2 above)
Comment (synthko) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java:139]: Yes, it makes much more sense as you pointed out in the Order model comment.

File: common/src/main/java/fi/hel/verkkokauppa/common/util/DateTimeUtil.java (hunk 3 above)
Comment (synthko) [reply common/src/main/java/fi/hel/verkkokauppa/common/util/DateTimeUtil.java:58]: Sure, I'll change this.

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java [hunk 4]
```diff
@@ -131,6 +131,20 @@ public Order createByParams(String namespace, String user) {
         return order;
     }
 
+    public Order createByParams(String namespace, String user, LocalDateTime lastValidPurchaseDateTime) {
+        LocalDateTime createdAt = DateTimeUtil.getFormattedDateTime();
+        String orderId = generateOrderId(namespace, user, createdAt);
+        Long incrementId = this.incrementId.generateOrderIncrementId();
+        Order order = new Order(orderId, namespace, user, createdAt, incrementId);
+        if (StringUtils.isNotEmpty(lastValidPurchaseDateTime.toString())) {
```
Comment (severikupari1) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java:139]: Can this cause NPO if lastValidPurchaseDateTime is null?

File: orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java [hunk 5]
```diff
@@ -645,4 +644,26 @@ void createOrderWithMerchantId() throws JsonProcessingException {
         orderRepository.save(order);
     }
 
+    @Test
+    @RunIfProfile(profile = "local")
+    public void CreateOrderWithLastValidPurchaseDateTime() {
```
Comment (severikupari1) [inline orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java:649]: You need to add test prefix to test functions ->
`testCreateOrderWithLastValidPurchaseDateTime`

File: orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java [hunk 6]
```diff
@@ -645,4 +644,26 @@ void createOrderWithMerchantId() throws JsonProcessingException {
...
+        Order order = orderRepository.findById(createOrderResponse.getOrder().getOrderId()).get();
+        LocalDateTime localDateTime = LocalDateTime.now();
+        order.setLastValidPurchaseDateTime(localDateTime);
+        orderRepository.save(order);
+        Order fetchedOrder = orderRepository.findById(order.getOrderId()).get();
+
+        /* Creating localDateTime with now() uses 6 digit precision with nanoseconds by default.
+           This is automatically rounded on save on db's side to 3 digit precision which causes assertion
+           error if values are asserted as is. Creating a formatter just for this felt excessive, thus
+           nanoseconds are rounded to 0 for "good-enough" assertion. */
```
Comment (severikupari1) [inline orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java:662]: This is good assertion and new thing to me, can make testing easier in some places where I had to create formatters.

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java (hunk 4 above)
Comment (martinldev) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java:139]: I do not think we need to even check the .toString() portion here. If we have an instance of lastValidPurchaseDateTime, then the toString() is not empty. So only **lastValidPurchaseDateTime != null** should be enough in this comparison.

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/api/OrderController.java [hunk 7]
```diff
@@ -75,9 +75,10 @@ public class OrderController {
 
     @GetMapping(value = "/order/create", produces = MediaType.APPLICATION_JSON_VALUE)
 	public ResponseEntity<OrderAggregateDto> createOrder(@RequestParam(value = "namespace") String namespace,
-                                                         @RequestParam(value = "user") String user) {
+                                                         @RequestParam(value = "user") String user,
+                                                         @RequestParam(required = false, value = "lastValidPurchaseDateTime") String lastValidPurchaseDateTime) {
         try {
-            Order order = orderService.createByParams(namespace, user);
+            Order order = orderService.createByParams(namespace, user, DateTimeUtil.fromFormattedDateTimeOptionalString(lastValidPurchaseDateTime));
```
Comment (martinldev) [inline orderapi/src/main/java/fi/hel/verkkokauppa/order/api/OrderController.java:81]: NPO is thrown here if "lastValidPurchaseDateTime" is null. Add a null check and do the formatting only if the value is provided.

**Review by martinldev — CHANGES_REQUESTED** [review]

File: orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java (hunk 5 above)
Comment (synthko) [reply orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java:649]: I can add the prefix.

Other methods seem to lack the prefix too in the class. While at it, shall I add the prefix to them too?

File: common/src/main/java/fi/hel/verkkokauppa/common/util/DateTimeUtil.java [hunk 8]
```diff
@@ -53,6 +55,15 @@ public static LocalDateTime fromFormattedDateTimeString(String localDateTime) {
         return LocalDateTime.parse(localDateTime, formatter);
     }
 
+    public static LocalDateTime fromFormattedDateTimeOptionalString(String localDateTime) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern(DATE_TIME_WITH_MILLISECONDS);
+
+        String caseStartDate = formatter.format(ZonedDateTime.now());
+        System.out.println(caseStartDate);
```
Comment (severikupari1) [inline common/src/main/java/fi/hel/verkkokauppa/common/util/DateTimeUtil.java:62]: delete :D

Comment (synthko) [reply common/src/main/java/fi/hel/verkkokauppa/common/util/DateTimeUtil.java:62]: Whoops, will do. 👍

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java (hunk 4 above)
Comment (synthko) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/service/order/OrderService.java:139]: Thanks both for the remarks. What Martin added seems to hold true, so I'll change this to be just a null check.

File: orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java (hunk 5 above)
Comment (synthko) [reply orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java:649]: As discussed in slack, I'll skip using the test -prefix and just change the case convention to camel.

File: orderapi/src/main/java/fi/hel/verkkokauppa/order/api/OrderController.java (hunk 7 above)
Comment (synthko) [reply orderapi/src/main/java/fi/hel/verkkokauppa/order/api/OrderController.java:81]: Good find - I'll add a check here 👍

**Comment by severikupari1** [comment]
Approved by me.

**Review by martinldev — APPROVED** [review]

File: orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java [hunk 9]
```diff
@@ -645,4 +644,49 @@ void createOrderWithMerchantId() throws JsonProcessingException {
...
+        Assertions.assertEquals(
+                localDateTime.withNano(0),
+                fetchedOrder.getLastValidPurchaseDateTime().withNano(0)
+        );
+    }
+    @Test
+    @RunIfProfile(profile = "local")
+    public void createOrderByParams() {
+        String namespace = "venepaikat";
```
[the remark this answers is not in the document]
Comment (synthko) [reply orderapi/src/test/java/fi/hel/verkkokauppa/order/service/order/OrderServiceTest.java:671]: Good tip, I'll add the not happy -path test to the code to cover that.

**Pull request merged**